### PR TITLE
Migrate to ESLint 9

### PR DESCRIPTION
**Pull request by tmns**
This PR upgrades `eslint` to `9.15.0`, which mainly required the following:
* Upgrade `prettier` to `3.4.1`, as Prettier 3 is needed for flat config compatibility. As par of this, also format all files with upgraded Prettier.
* Migrate the existing config to the now standard ["flat" config](https://eslint.org/docs/latest/use/configure/migration-guide#start-using-flat-config-files) API, requiring a rename to `eslint.config.js`. Note, React-specific configs / plugins weren't carried over to the new config, as they didn't seem necessary for this codebase.
* Remove `.eslintignore`, as ignores are now required to be defined within the config itself.
* Fix up some lint errors that began popping up after the upgrade.

**Testing**
It'd be good to double check linting works in your editor & terminal.

**Resources**
* [ESLint Docs](https://eslint.org/docs/latest/use/getting-started)
* [`typescript-eslint` Docs](https://typescript-eslint.io/getting-started/)

Notion card [here](https://www.notion.so/vertexprotocol/Upgrade-ESLint-Prettier-in-SDK-03b312d3e62949b298e6f946262062f0?pvs=4).

### CODE REVIEW

File: eslint.config.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+const tseslint = require('typescript-eslint');
+const eslint = require('@eslint/js');
+const prettierPluginRecommended = require('eslint-plugin-prettier/recommended');
+const typescriptEslint = require('@typescript-eslint/eslint-plugin');
+const globals = require('globals');
+
+module.exports = tseslint.config(
+  eslint.configs.recommended,
+  tseslint.configs.recommended,
+  prettierPluginRecommended,
```
Comment (tmns) [inline eslint.config.js:10]: By importing `eslint...prettier/recommended` and including it here, both the plugin and config get set up in one go, meaning we don't need to also also add the plugin to `plugins` below. See [here](https://github.com/prettier/eslint-plugin-prettier?tab=readme-ov-file#configuration-new-eslintconfigjs) for more info.

File: eslint.config.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
```
Comment (tmns) [inline eslint.config.js:10]: By importing `eslint...prettier/recommended` and including it here, both the config and plugin get set up in one go, meaning we don't have to also include the plugin in `plugins` below. See [here](https://github.com/prettier/eslint-plugin-prettier?tab=readme-ov-file#configuration-new-eslintconfigjs) for more info.

File: eslint.config.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
+const tseslint = require('typescript-eslint');
+const eslint = require('@eslint/js');
+const prettierPluginRecommended = require('eslint-plugin-prettier/recommended');
+const typescriptEslint = require('@typescript-eslint/eslint-plugin');
+const globals = require('globals');
+
+module.exports = tseslint.config(
+  eslint.configs.recommended,
+  tseslint.configs.recommended,
```
Comment (tmns) [inline eslint.config.js:9]: I initially tried `tseslint.configs.recommendedTypeChecked` here instead but there were a fare amount of issues that I wouldn't have the time nor context to adequately sort out. But you all may wanna look into enabling it at some point.

See docs [here](https://typescript-eslint.io/getting-started/typed-linting).

File: eslint.config.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      'prettier/prettier': 'error',
+    },
+  },
+  {
+    ignores: [
+      '**/.nx/',
+      'packages/contracts/src/typechain-types',
+      '**/node_modules/',
+      '**/.next',
+      '**/dist',
```
Comment (tmns) [inline eslint.config.js:58]: From the looks of it, only the above are necessary to carry over from the `.eslintignore` file. But feel free to add more if I'm missing something.

**Comment by frankfka** [comment]
Thanks Methias! Ill dig into some of the comments you raised here tomorrow

File: eslint.config.js (hunk 3 above)
Comment (frankfka) [reply eslint.config.js:9]: Just to clarify, you got the config working, but that resulted in quite a few linter errors?

**Comment by frankfka** [comment]
Just realized, Cedric had a past PR for this as well, not sure if he found / exposed anything that was missing here: https://github.com/vertex-protocol/vertex-typescript-sdk/pull/224/files

File: eslint.config.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  eslint.configs.recommended,
+  tseslint.configs.recommended,
+  prettierPluginRecommended,
+  {
+    languageOptions: {
+      // Ensures we can use global vars like `__dirname`.
+      globals: { ...globals.node },
+    },
+    rules: {
+      // Allows us to use `require` in `js` files.
```
[the remark this answers is not in the document]
Comment (tmns) [reply eslint.config.js:17]: Just this one! But this file also gets linted.

File: eslint.config.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      'no-constant-condition': 'off',
+      'prettier/prettier': 'error',
+    },
+  },
+  {
+    ignores: [
+      '**/.nx/',
+      'packages/contracts/src/typechain-types',
+      '**/node_modules/',
+      '**/.next',
```
[the remark this answers is not in the document]
Comment (tmns) [reply eslint.config.js:57]: Ah of course.

File: eslint.config.js (hunk 3 above)
Comment (tmns) [reply eslint.config.js:9]: Yep, that's right. They didn't seem that serious or difficult to solve though, a lot of them were seemingly unnecessary `await`s and implicit use of `any`s.

**Comment by tmns** [comment]
> Just realized, Cedric had a past PR for this as well, not sure if he found / exposed anything that was missing here: https://github.com/vertex-protocol/vertex-typescript-sdk/pull/224/files

Ah nice, thanks. I took a look through it and didn't see anything not covered here.

File: eslint.config.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (frankfka) [reply eslint.config.js:17]: is there a good way to set an exception for this file?

**Comment by frankfka** [comment]
Looks like ESLint highlighted a bunch of valid concerns 👍🏻 addressed via a commit

File: eslint.config.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tmns) [reply eslint.config.js]: Yep, I also found that `globals` is only needed for the config file, so adjusted that as well.

**Comment by frankfka** [comment]
Ill try some of these changes locally with the frontend first

**Pull request merged**